crates/starknet-devnet-server/src/api/json_rpc/origin_forwarder.rs

```diff
@@ -0,0 +1,85 @@
+use axum::http::request;
+use hyper::client::HttpConnector;
+
+use crate::rpc_core::error::RpcError;
+use crate::rpc_core::request::RpcMethodCall;
+use crate::rpc_core::response::{ResponseResult, RpcResponse};
+
+/// Used for forwarding requests to origin in case of:
+/// - BlockNotFound
+/// - TransactionNotFound
+/// - NoStateAtBlock
+/// - ClassHashNotFound
+///
+/// Basic contract-wise interaction is handled by `BlockingOriginReader`
+#[derive(Clone)]
+pub struct OriginForwarder {
+    client: hyper::Client<HttpConnector>,
+    url: hyper::Uri,
+    block_number: u64,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum OriginInteractionError {}
+
+impl OriginForwarder {
+    pub fn new(url: hyper::Uri, block_number: u64) -> Self {
+        Self { client: hyper::Client::new(), url, block_number }
+    }
+
+    /// In case block tag "pending" or "latest" is a part of the request, it is replaced with the
+    /// numeric block id of the forked block. Both JSON-RPC 1 and 2 semantics is covered
+    fn clone_call_with_origin_block_id(&self, rpc_call: &RpcMethodCall) -> RpcMethodCall {
+        let mut new_rpc_call = rpc_call.clone();
+        let origin_block_id = serde_json::json!({ "block_number": self.block_number });
+
+        match new_rpc_call.params {
+            crate::rpc_core::request::RequestParams::None => (),
+            crate::rpc_core::request::RequestParams::Array(ref mut params) => {
+                for param in params.iter_mut() {
+                    if let Some("latest" | "pending") = param.as_str() {
+                        *param = origin_block_id;
```

Comment:
in json-rpc 1.0 when params is an array, this is the format of the request {"method": "handleMessage", "params": ["user1", "we were just talking"], "id":null} https://www.jsonrpc.org/specification_v1#a4.CommunicationExamples
I think origin_block_id will replace "latest" or "pending" with the wrong format. It will result in "params": ["\"block_number: 12312\"", "next param"]

Reply:
I'll try adding unit tests to this part of the code